[Quant][PT2E] Enable linear and linear-unary post-op quant recipe for x86 inductor quantizer

Xia-Weiwen opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #106935
* #106934
* #106782
* __->__ #106781
* #105818

**Summary**
Add linear and linear-unary post-op quantization recipe to x86 inductor quantizer. For PT2E with Inductor. With this, the quantization path will add `quant-dequant` pattern for linear and linear-unary post op.

**Test plan**
python test/test_quantization.py -k test_linear_with_quantizer_api
python test/test_quantization.py -k test_linear_unary_with_quantizer_api

Xia-Weiwen committed:
[Quant][PT2E] Enable linear and linear-unary post-op quant recipe for x86 inductor quantizer

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant][PT2E] Enable linear and linear-unary post-op quant recipe for x86 inductor quantizer"




[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant][PT2E] Enable linear and linear-unary post-op quant recipe for x86 inductor quantizer"




[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant][PT2E] Enable linear and linear-unary post-op quant recipe for x86 inductor quantizer"


**Summary**
Add linear and linear-unary post-op quantization recipe to x86 inductor quantizer. For PT2E with Inductor. With this, the quantization path will add `quant-dequant` pattern for linear and linear-unary post op.

**Test plan**
python test/test_quantization.py -k test_linear_with_quantizer_api
python test/test_quantization.py -k test_linear_unary_with_quantizer_api


[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant][PT2E] Enable linear and linear-unary post-op quant recipe for x86 inductor quantizer"


**Summary**
Add linear and linear-unary post-op quantization recipe to x86 inductor quantizer. For PT2E with Inductor. With this, the quantization path will add `quant-dequant` pattern for linear and linear-unary post op.

**Test plan**
python test/test_quantization.py -k test_linear_with_quantizer_api
python test/test_quantization.py -k test_linear_unary_with_quantizer_api


[ghstack-poisoned]

jgong5 commented on test/quantization/pt2e/test_x86inductor_quantizer.py:
```diff
@@ -153,6 +153,26 @@ def forward(self, x):
+        def __init__(self, use_bias, postop, inplace_postop) -> None:
+            super().__init__()
+            self.linear = nn.Linear(4, 4, bias=use_bias)
+            if postop is nn.Tanh:
```
Do we have `nn.Tanh` postop covered by test?

Xia-Weiwen replied:
Thanks. It's removed.

Xia-Weiwen committed:
Update on "[Quant][PT2E] Enable linear and linear-unary post-op quant recipe for x86 inductor quantizer"


**Summary**
Add linear and linear-unary post-op quantization recipe to x86 inductor quantizer. For PT2E with Inductor. With this, the quantization path will add `quant-dequant` pattern for linear and linear-unary post op.

**Test plan**
python test/test_quantization.py -k test_linear_with_quantizer_api
python test/test_quantization.py -k test_linear_unary_with_quantizer_api


[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant][PT2E] Enable linear and linear-unary post-op quant recipe for x86 inductor quantizer"


**Summary**
Add linear and linear-unary post-op quantization recipe to x86 inductor quantizer. For PT2E with Inductor. With this, the quantization path will add `quant-dequant` pattern for linear and linear-unary post op.

**Test plan**
python test/test_quantization.py -k test_linear_with_quantizer_api
python test/test_quantization.py -k test_linear_unary_with_quantizer_api


[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant][PT2E] Enable linear and linear-unary post-op quant recipe for x86 inductor quantizer"


**Summary**
Add linear and linear-unary post-op quantization recipe to x86 inductor quantizer. For PT2E with Inductor. With this, the quantization path will add `quant-dequant` pattern for linear and linear-unary post op.

**Test plan**
python test/test_quantization.py -k test_linear_with_quantizer_api
python test/test_quantization.py -k test_linear_unary_with_quantizer_api


[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant][PT2E] Enable linear and linear-unary post-op quant recipe for x86 inductor quantizer"


**Summary**
Add linear and linear-unary post-op quantization recipe to x86 inductor quantizer. For PT2E with Inductor. With this, the quantization path will add `quant-dequant` pattern for linear and linear-unary post op.

**Test plan**
python test/test_quantization.py -k test_linear_with_quantizer_api
python test/test_quantization.py -k test_linear_unary_with_quantizer_api


[ghstack-poisoned]

leslie-fang-intel commented:
@jerryzh168 Could you also kindly help to take a look of this PR?

jerryzh168 commented on test/quantization/pt2e/test_x86inductor_quantizer.py:
```diff
@@ -400,3 +417,80 @@ def test_conv2d_serials_binary_unary_with_quantizer_api(self):
             )
+
+    @skipIfNoX86
+    def test_linear_with_quantizer_api(self):
```
nit: you can remove `_with_quantizer_api` now

Xia-Weiwen replied:
Thanks. It's removed.

jerryzh168 commented on test/quantization/pt2e/test_x86inductor_quantizer.py:
```diff
@@ -400,3 +417,80 @@ def test_conv2d_serials_binary_unary_with_quantizer_api(self):
+                )
+
+    @skipIfNoX86
+    def test_linear_unary_with_quantizer_api(self):
```
same here

Xia-Weiwen replied:
Thanks. It's removed.

jerryzh168 commented on test/quantization/pt2e/test_x86inductor_quantizer.py:
```diff
@@ -400,3 +417,80 @@ def test_conv2d_serials_binary_unary_with_quantizer_api(self):
+                node_list = [
+                    torch.ops.quantized_decomposed.quantize_per_tensor.default,
+                    torch.ops.quantized_decomposed.dequantize_per_tensor.default,
+                    torch.ops.aten.addmm.default if use_bias else torch.ops.aten.mm.default,
```
if you don't need to land this before branch cut, I think you could switch to `capture_pre_autograd_graph` for capture and you will see a single aten.linear op instead

Xia-Weiwen replied:
Hi @jerryzh168. Thanks for the suggestion. We hope to land this before code freeze. We feel difficult to switch to new API as code freeze is near. Is it OK to land this first then we switch to new API later?

jerryzh168 replied:
sure, that's fine

jerryzh168 commented on torch/ao/quantization/quantizer/x86_inductor_quantizer.py:
```diff
@@ -199,6 +199,52 @@ def _annotate_conv_node_helper(
+        assert isinstance(input_node, Node)
+        input_qspec_map[input_node] = get_input_act_qspec(quantization_config)
+
+        t_node = linear_node.args[weight_index]
+        assert isinstance(t_node, Node)
```
I think switching to use the new capture will be easier, this matching is not covering all use cases since there could be sym_size ops etc. I think ideally you can wait until https://github.com/pytorch/pytorch/pull/107872 and change the capture method

Xia-Weiwen replied:
Same here. Is it OK to land this first then we switch to new API later? Thanks.

Xia-Weiwen committed:
Update on "[Quant][PT2E] Enable linear and linear-unary post-op quant recipe for x86 inductor quantizer"


**Summary**
Add linear and linear-unary post-op quantization recipe to x86 inductor quantizer. For PT2E with Inductor. With this, the quantization path will add `quant-dequant` pattern for linear and linear-unary post op.

**Test plan**
python test/test_quantization.py -k test_linear_with_quantizer_api
python test/test_quantization.py -k test_linear_unary_with_quantizer_api


[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant][PT2E] Enable linear and linear-unary post-op quant recipe for x86 inductor quantizer"


**Summary**
Add linear and linear-unary post-op quantization recipe to x86 inductor quantizer. For PT2E with Inductor. With this, the quantization path will add `quant-dequant` pattern for linear and linear-unary post op.

**Test plan**
python test/test_quantization.py -k test_linear_with_quantizer_api
python test/test_quantization.py -k test_linear_unary_with_quantizer_api


[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant][PT2E] Enable linear and linear-unary post-op quant recipe for x86 inductor quantizer"


**Summary**
Add linear and linear-unary post-op quantization recipe to x86 inductor quantizer. For PT2E with Inductor. With this, the quantization path will add `quant-dequant` pattern for linear and linear-unary post op.

**Test plan**
python test/test_quantization.py -k test_linear_with_quantizer_api
python test/test_quantization.py -k test_linear_unary_with_quantizer_api


[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant][PT2E] Enable linear and linear-unary post-op quant recipe for x86 inductor quantizer"


**Summary**
Add linear and linear-unary post-op quantization recipe to x86 inductor quantizer. For PT2E with Inductor. With this, the quantization path will add `quant-dequant` pattern for linear and linear-unary post op.

**Test plan**
python test/test_quantization.py -k test_linear_with_quantizer_api
python test/test_quantization.py -k test_linear_unary_with_quantizer_api


[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant][PT2E] Enable linear and linear-unary post-op quant recipe for x86 inductor quantizer"


**Summary**
Add linear and linear-unary post-op quantization recipe to x86 inductor quantizer. For PT2E with Inductor. With this, the quantization path will add `quant-dequant` pattern for linear and linear-unary post op.

**Test plan**
python test/test_quantization.py -k test_linear_with_quantizer_api
python test/test_quantization.py -k test_linear_unary_with_quantizer_api


[ghstack-poisoned]

Xia-Weiwen commented:
@pytorchbot merge
